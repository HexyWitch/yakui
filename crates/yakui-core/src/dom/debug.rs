use std::fmt;

use super::{Dom, DomInner};

impl fmt::Debug for Dom {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.inner.try_borrow() {
            Ok(inner) => f
                .debug_struct("Dom")
                .field("roots", &inner.roots)
                .field("nodes", &ViewTree(&inner))
                .finish(),

            Err(_) => f.debug_struct("Dom (Locked)").finish(),
        }
    }
}

struct ViewTree<'a>(&'a DomInner);

impl<'a> fmt::Debug for ViewTree<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let dom = &self.0;
        let iter = dom.nodes.iter().map(|(index, node)| {
            let debug = node.widget.as_debug();

            let children: Vec<_> = node.children.iter().map(|index| index.slot()).collect();

            format!("{}: {debug:?}, children: {:?}", index.slot(), children)
        });

        f.debug_list().entries(iter).finish()
    }
}
