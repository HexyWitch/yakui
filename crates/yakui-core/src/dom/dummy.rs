use crate::Widget;

/// Placeholder widget used internally to emplace a component without
/// initializing it yet.
#[derive(Debug)]
pub(crate) struct DummyWidget;

impl Widget for DummyWidget {
    type Props = ();
    type Response = ();

    #[inline]
    fn new(_props: Self::Props) -> Self {
        Self
    }

    #[inline]
    fn update(&mut self, _props: Self::Props) {}

    #[inline]
    fn respond(&mut self) {}
}
