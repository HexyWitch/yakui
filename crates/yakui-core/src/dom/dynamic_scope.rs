use std::any::Any;
use std::any::TypeId;
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

#[derive(Default)]
pub struct DynamicScope {
    inner: RefCell<Inner>,
}

#[derive(Default)]
struct Inner {
    scopes: Vec<Scope>,
    stack: Vec<usize>,
}

#[derive(Clone, Default)]
pub struct Scope {
    storage: HashMap<TypeId, Rc<dyn Any>>,
}

impl DynamicScope {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn clear(&self) {
        let mut inner = self.inner.borrow_mut();
        inner.scopes.clear();
        inner.stack.clear();
    }

    pub fn get<T: 'static>(&self, scope: usize) -> Option<Rc<T>> {
        let inner = self.inner.borrow();

        let scope = inner.scopes.get(scope)?;
        let item = scope.storage.get(&TypeId::of::<T>())?;
        Some(Rc::downcast(item.clone()).unwrap())
    }

    pub fn current_scope(&self) -> Option<usize> {
        let inner = self.inner.borrow();
        inner.stack.last().copied()
    }

    pub fn push_scope(&self) -> usize {
        let mut inner = self.inner.borrow_mut();
        let mut new_scope = Scope::default();

        if let Some(parent) = inner.stack.last().and_then(|&i| inner.scopes.get(i)) {
            new_scope = parent.clone();
        }

        let index = inner.scopes.len();
        inner.scopes.push(new_scope);
        inner.stack.push(index);

        index
    }

    #[track_caller]
    pub fn pop_scope(&self) {
        let mut inner = self.inner.borrow_mut();

        assert!(
            inner.stack.pop().is_some(),
            "cannot pop_scope when the stack is empty"
        );
    }

    pub fn write_item<T: 'static>(&self, value: T) {
        let mut inner = self.inner.borrow_mut();
        let inner = &mut *inner;

        let scope = inner
            .stack
            .last()
            .and_then(|&i| inner.scopes.get_mut(i))
            .unwrap_or_else(|| panic!("cannot write_item when the stack is empty"));

        scope.storage.insert(TypeId::of::<T>(), Rc::new(value));
    }
}
