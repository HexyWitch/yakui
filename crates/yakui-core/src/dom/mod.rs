mod debug;

use std::any::{Any, TypeId};
use std::collections::VecDeque;

use anymap::AnyMap;
use thunderdome::{Arena, Index};

use crate::component::{Component, DummyComponent, ErasedComponent};

pub struct Dom {
    tree: Arena<DomNode>,
    roots: Vec<Index>,

    color: bool,
    stack: Vec<Index>,
    build_index: usize,

    global_state: AnyMap,
}

pub struct DomNode {
    pub component: Box<dyn ErasedComponent>,
    pub children: Vec<Index>,
    build_index: usize,
    color: bool,
}

impl DomNode {
    pub fn set_color(&mut self, color: bool) {
        if self.color != color {
            self.build_index = 0;
        }

        self.color = color;
    }
}

impl Dom {
    pub fn new() -> Self {
        Self {
            tree: Arena::new(),
            roots: Vec::new(),

            color: false,
            stack: Vec::new(),
            build_index: 0,

            global_state: AnyMap::new(),
        }
    }

    pub fn start(&mut self) {
        self.color = !self.color;
        self.build_index = 0;
    }

    pub fn do_component<T: Component>(&mut self, props: T::Props) -> T::Response {
        let index = self.begin_component::<T>(props);
        self.end_component::<T>(index)
    }

    pub fn begin_component<T: Component>(&mut self, props: T::Props) -> Index {
        let parent = self.stack.last();

        let index = match parent {
            Some(&parent_index) => {
                let parent = self.tree.get_mut(parent_index).unwrap();
                parent.set_color(self.color);

                if parent.build_index < parent.children.len() {
                    let index = parent.children[parent.build_index];
                    parent.build_index += 1;
                    index
                } else {
                    let index = self.tree.insert(DomNode {
                        component: Box::new(DummyComponent),
                        children: Vec::new(),
                        build_index: 0,
                        color: self.color,
                    });

                    let parent = self.tree.get_mut(parent_index).unwrap();
                    parent.children.push(index);
                    parent.build_index += 1;
                    index
                }
            }
            None => {
                if self.build_index < self.roots.len() {
                    let index = self.roots[self.build_index];
                    self.build_index += 1;
                    index
                } else {
                    let index = self.tree.insert(DomNode {
                        component: Box::new(DummyComponent),
                        children: Vec::new(),
                        build_index: 0,
                        color: self.color,
                    });
                    self.roots.push(index);
                    self.build_index += 1;
                    index
                }
            }
        };

        self.stack.push(index);

        let node = self.tree.get_mut(index).unwrap();
        if node.component.as_ref().type_id() == TypeId::of::<T>() {
            let component = node.component.downcast_mut::<T>().unwrap();
            component.update(props);
        } else {
            node.component = Box::new(T::new(index, props));
        }

        index
    }

    pub fn end_component<T: Component>(&mut self, index: Index) -> T::Response {
        let old_top = self.stack.pop().unwrap_or_else(|| {
            panic!("Cannot end_component without an in-progress component.");
        });

        assert!(
            index == old_top,
            "Dom::end_component did not match the input component."
        );

        self.trim_children(index);

        let node = self.tree.get_mut(index).unwrap();

        node.component
            .as_mut()
            .downcast_mut::<T>()
            .unwrap()
            .respond()
    }

    pub fn get_global_state<T: Any>(&self) -> Option<&T> {
        self.global_state.get::<T>()
    }

    pub fn set_global_state<T: Any>(&mut self, value: T) -> Option<T> {
        self.global_state.insert::<T>(value)
    }

    pub fn get_global_state_or_insert_with<T: Any, F: FnOnce() -> T>(&mut self, init: F) -> &mut T {
        self.global_state.entry::<T>().or_insert_with(init)
    }

    /// Remove children from the given node that weren't present in the latest
    /// traversal through the tree.
    fn trim_children(&mut self, index: Index) {
        let node = self.tree.get_mut(index).unwrap();

        if node.build_index < node.children.len() {
            let mut queue = VecDeque::new();
            let to_drop = &node.children[node.build_index..];
            queue.extend(to_drop);

            node.children.truncate(self.build_index);

            while let Some(index) = queue.pop_front() {
                let node = self.tree.remove(index).unwrap();
                queue.extend(node.children);
            }
        }
    }

    pub fn roots(&self) -> &[Index] {
        self.roots.as_slice()
    }

    pub fn get(&self, index: Index) -> Option<&DomNode> {
        self.tree.get(index)
    }

    pub fn get_mut(&mut self, index: Index) -> Option<&mut DomNode> {
        self.tree.get_mut(index)
    }
}
