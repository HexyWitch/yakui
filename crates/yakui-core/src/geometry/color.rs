use glam::Vec3;

/// An sRGB color.
#[derive(Debug, Clone, Copy, PartialEq)]
#[allow(missing_docs)]
pub struct Color3 {
    pub r: u8,
    pub g: u8,
    pub b: u8,
}

impl Color3 {
    /// Create a new `Color3`.
    pub const fn rgb(r: u8, g: u8, b: u8) -> Self {
        Self { r, g, b }
    }

    /// Create a color from a number, intended to be written as a hex literal.
    ///
    /// ```rust
    /// # use yakui_core::geometry::Color3;
    /// let orange = Color3::hex(0xff6137);
    /// assert_eq!(orange, Color3::rgb(255, 97, 55));
    /// ```
    pub const fn hex(value: u32) -> Self {
        let r = ((value >> 16) & 255) as u8;
        let g = ((value >> 8) & 255) as u8;
        let b = (value & 255) as u8;

        Self { r, g, b }
    }

    /// Create a new `Color3` from a linear RGB color.
    pub fn from_linear(value: Vec3) -> Self {
        let linear = palette::LinSrgb::new(value.x, value.y, value.z);
        let (r, g, b) = palette::Srgb::from_linear(linear)
            .into_format::<u8>()
            .into_components();

        Self::rgb(r, g, b)
    }

    /// Brighten or darken the color by multiplying it by a factor.
    ///
    /// Values greater than 1 will lighten the color, values smaller than 1 will
    /// darken it.
    ///
    /// This operation is gamma-correct.
    pub fn adjust(&self, factor: f32) -> Self {
        Self::from_linear(self.to_linear() * factor)
    }

    /// Convert this color to a linear RGB color.
    pub fn to_linear(&self) -> Vec3 {
        palette::Srgb::new(self.r, self.g, self.b)
            .into_format::<f32>()
            .into_linear()
            .into_components()
            .into()
    }
}

macro_rules! builtin_colors {
    ($($name:ident ($($color:literal),*),)*) => {
        #[allow(missing_docs)]
        impl Color3 {
            $(
                pub const $name: Self = Self::rgb($($color),*);
            )*
        }
    }
}

builtin_colors! {
    RED(255, 0, 0),
    GREEN(0, 255, 0),
    BLUE(0, 0, 255),
    YELLOW(255, 255, 0),
    CYAN(0, 255, 255),
    FUCHSIA(255, 0, 255),
    WHITE(255, 255, 255),
    GRAY(127, 127, 127),
    BLACK(0, 0, 0),

    CORNFLOWER_BLUE(100, 149, 237),
    REBECCA_PURPLE(102, 51, 153),
}

impl From<[u8; 3]> for Color3 {
    #[inline]
    fn from(value: [u8; 3]) -> Self {
        Self {
            r: value[0],
            g: value[1],
            b: value[2],
        }
    }
}
