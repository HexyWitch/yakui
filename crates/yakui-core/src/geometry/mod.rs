//! Defines common geometry types used by yakui.

mod color;
mod constraints;
mod dim;
mod rect;
mod urect;

#[doc(no_inline)]
pub use glam::{UVec2, Vec2, Vec4};

pub use self::color::*;
pub use self::constraints::*;
pub use self::dim::*;
pub use self::rect::*;
pub use self::urect::*;

/// Defines how a flexible container should size its children.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FlexFit {
    /// The container forces its children to stretch to its size.
    Tight,

    /// The container lets the child have any size that fits within the
    /// container.
    Loose,
}
