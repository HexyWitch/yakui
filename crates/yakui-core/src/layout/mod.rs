use std::collections::VecDeque;

use glam::Vec2;
use thunderdome::{Arena, Index};

use crate::dom::Dom;
use crate::geometry::{Constraints, Rect};

#[derive(Debug)]
pub struct LayoutDom {
    pub viewport: Rect,
    nodes: Arena<LayoutDomNode>,
}

#[derive(Debug)]
pub struct LayoutDomNode {
    pub rect: Rect,
}

impl LayoutDom {
    pub fn new() -> Self {
        Self {
            viewport: Rect::ZERO,
            nodes: Arena::new(),
        }
    }

    pub fn clear(&mut self) {
        self.nodes.clear();
    }

    pub fn get(&self, index: Index) -> Option<&LayoutDomNode> {
        self.nodes.get(index)
    }

    pub fn get_mut(&mut self, index: Index) -> Option<&mut LayoutDomNode> {
        self.nodes.get_mut(index)
    }

    pub fn calculate_all(&mut self, dom: &Dom) {
        let constraints = Constraints {
            min: Vec2::ZERO,
            max: self.viewport.size(),
        };

        for &index in dom.roots() {
            self.calculate(dom, index, constraints);
        }

        self.resolve_positions(dom);
    }

    pub fn calculate(&mut self, dom: &Dom, index: Index, constraints: Constraints) -> Vec2 {
        let dom_node = dom.get(index).unwrap();
        let size = dom_node.widget.layout(dom, self, constraints);
        self.nodes.insert_at(
            index,
            LayoutDomNode {
                rect: Rect::from_pos_size(Vec2::ZERO, size),
            },
        );
        size
    }

    pub fn set_pos(&mut self, index: Index, pos: Vec2) {
        if let Some(node) = self.nodes.get_mut(index) {
            node.rect.set_pos(pos);
        }
    }

    fn resolve_positions(&mut self, dom: &Dom) {
        let mut queue = VecDeque::new();

        queue.extend(dom.roots().iter().map(|&index| (index, Vec2::ZERO)));

        while let Some((index, parent_pos)) = queue.pop_front() {
            if let Some(layout_node) = self.nodes.get_mut(index) {
                let node = dom.get(index).unwrap();
                layout_node
                    .rect
                    .set_pos(layout_node.rect.pos() + parent_pos);

                queue.extend(
                    node.children
                        .iter()
                        .map(|&index| (index, layout_node.rect.pos())),
                );
            }
        }
    }
}
