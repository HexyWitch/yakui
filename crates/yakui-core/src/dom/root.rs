use crate::layout::LayoutDom;
use crate::paint::PaintDom;
use crate::{Constraints, Widget};

use super::Dom;

#[derive(Debug)]
pub struct RootWidget;

impl Widget for RootWidget {
    type Props = ();
    type Response = ();

    fn new(_props: Self::Props) -> Self {
        Self
    }

    fn update(&mut self, _props: Self::Props) {}
    fn respond(&mut self) -> Self::Response {}

    fn layout(&self, dom: &Dom, layout: &mut LayoutDom, constraints: Constraints) -> glam::Vec2 {
        let node = dom.get_current();

        for &child in &node.children {
            layout.calculate(dom, child, constraints);
        }

        constraints.max
    }

    fn paint(&self, dom: &Dom, layout: &LayoutDom, paint: &mut PaintDom) {
        let node = dom.get_current();

        for &child in &node.children {
            let child = dom.get(child).unwrap();
            child.widget.paint(dom, layout, paint);
        }
    }
}
