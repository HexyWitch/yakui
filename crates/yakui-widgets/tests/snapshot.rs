use std::collections::VecDeque;
use std::fmt::Write as _;

use yakui_core::dom::Dom;
use yakui_core::geometry::{Color3, Rect, Vec2};
use yakui_core::layout::LayoutDom;
use yakui_core::Alignment;

#[test]
fn colored_box() {
    Test::new("colored_box").run(|| {
        yakui_widgets::colored_box(Color3::WHITE, [50.0, 50.0]);
    });
}

#[test]
fn column() {
    Test::new("column").run(|| {
        yakui_widgets::column(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn row() {
    Test::new("row").run(|| {
        yakui_widgets::row(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn align_center() {
    Test::new("align_center").run(|| {
        yakui_widgets::center(|| {
            rect_50x50();
        });
    });
}

#[test]
fn align_bottom_right() {
    Test::new("align_bottom_right").run(|| {
        yakui_widgets::align(Alignment::BOTTOM_RIGHT, || {
            rect_50x50();
        });
    });
}

fn rect(w: f32, h: f32) {
    yakui_widgets::colored_box(Color3::WHITE, [w, h]);
}

fn rect_50x50() {
    rect(50.0, 50.0);
}

struct Test {
    name: &'static str,
    viewport: Rect,
}

impl Test {
    fn new(name: &'static str) -> Self {
        Self {
            name,
            viewport: Rect::from_pos_size(Vec2::ZERO, Vec2::new(1000.0, 1000.0)),
        }
    }

    fn run<F: FnOnce()>(self, callback: F) {
        let mut state = yakui_core::State::new();
        state.set_unscaled_viewport(self.viewport);
        state.start();
        callback();
        state.finish();

        let dom = state.dom();
        let layout = state.layout_dom();

        insta::assert_snapshot!(self.name, view(dom, layout));
    }
}

fn view(dom: &Dom, layout: &LayoutDom) -> String {
    let mut output = String::new();
    let mut to_visit = VecDeque::new();

    let root = dom.get(dom.root()).unwrap();
    to_visit.extend(root.children.iter().rev().map(|&id| (id, 0)));

    while let Some((id, depth)) = to_visit.pop_back() {
        let node = dom.get(id).unwrap();
        let indent = "  ".repeat(depth);
        let name = trim_name(node.widget.type_name());

        let layout_info = match layout.get(id) {
            Some(info) => {
                let [x, y] = info.rect.pos().to_array();
                let [w, h] = info.rect.size().to_array();

                format!("pos({x}, {y}) size({w}, {h})")
            }
            None => "(no layout)".to_owned(),
        };

        writeln!(output, "{indent}- {name} {layout_info}").unwrap();

        to_visit.extend(node.children.iter().rev().map(|&id| (id, depth + 1)));
    }

    output
}

/// Trim modules off of the name of a Rust type name.
fn trim_name(name: &'static str) -> &'static str {
    name.rsplit("::").next().unwrap()
}
