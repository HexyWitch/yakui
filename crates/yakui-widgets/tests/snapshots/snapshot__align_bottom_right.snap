---
source: crates/yakui-widgets/tests/snapshot.rs
assertion_line: 87
expression: "view(dom, layout)"
---
- AlignWidget pos(0, 0) size(1000, 1000)
  - ColoredBoxWidget pos(950, 950) size(50, 50)

